sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -650,15 +646,15 @@ object FoldablePropagation extends Rule[LogicalPlan] {
 
     if (foldableExprSet.isEmpty) {
       plan
-    } else if (hasCommandOrUnion(plan)) {
-      replaceInSubplan(plan)
     } else {
       val foldableMap = AttributeMap(foldableExprSet.toSeq.map(_.asInstanceOf[Alias])
         .map(a => (a.toAttribute, a.child)))
 
-      CleanupAliases(plan.transformAllExpressions {
-        case a : AttributeReference if foldableMap.contains(a) =>
-          Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)()
+      CleanupAliases(plan.transformUp {
+        case p: LogicalPlan if !hasCommandOrUnion(p) => p.transformAllExpressions {
+          case a: AttributeReference if foldableMap.contains(a) =>
+            Alias(foldableMap.getOrElse(a.toAttribute, a), a.name)()
+        }
```

Comment:
By the way, in the above code using `stop` flag, if earlier sibling has `Union`, the later sibling is disabled. So, it's not safe. It's just an example.